base/compiler/abstractinterpretation.jl

```diff
@@ -1805,7 +1805,7 @@ function abstract_eval_global(M::Module, s::Symbol)
     if isdefined(M,s) && isconst(M,s)
         return Const(getfield(M,s))
     end
-    return Any
+    return Core.get_binding_type(M, s)
```

Comment:
No, this should never throw. Do you mean it's impure because the binding type can be redefined? That's the same issue we have with constants already.

Reply:
It is impure because it can start without a binding type, and be added later.

Reply:
Yes, I agree that it's a little iffy. If we don't mark it as pure though, the `convert` step will always be dynamic. Like I said, my thinking was that this is similar to how we assume purity of `const` values when resolving `GlobalRef`s.

Reply:
yes, we need to model it more precisely, so that it knows whether the global has a type and/or value assigned yet

Reply:
potentially we can make it syntactic, so that you know, as soon as the method is defined, whether there will be a type already given or not on the global

Reply:
(which would make it pure)

Reply:
> potentially we can make it syntactic, so that you know, as soon as the method is defined, whether there will be a type already given or not on the global

Could you elaborate on that? Do you mean we basically make a substitution right after lowering?

Reply:
If I understand you correctly, wouldn't this approach fail in cases like:

```julia
begin
    global x::Int
    x = 1.0
end
```

Or maybe we could add back some logic I had here previously to analyze cases like these in the frontend? 🤔

Reply:
yes, doing inference on code like that is already buggy for that reason

We could make the substitution early, but I actually mean that a `global` decl should simultaneously set the type (to Any), except where the global already has a type. This would prohibit code ordering like:
```julia
f() = (global x; x += 1)
x::Int = 3
```

But I think it is probably better for the global type declarations to appear first (before their usage) anyways

Reply:
Ok, that's interesting. You mean, we would fix it to `Any` anytime there's a write to a global? Or should `f() = (global x; x)` already count as fixing the type to `Any`?

Reply:
`f() = (global x; x)` already means the same as `global x; f() = x`, so both would fix the type to Any.

Reply:
`x` is never written to, only read from in that example though. I thought we'd only fix that type if the function actually writes to the binding, since we only need the binding type if we actually do an assignment.

Reply:
Ok, I have made that change now, but note that the type is only fixed if `x` is actually written to.